### ISSUE TITLE
WooPay reduces stock even if the checkout does not finish

### ISSUE THREAD

**Issue by ncsramona** [Automattic/woocommerce-payments#7257]
### Describe the bug
When going to the WooPay checkout page and going back to cart, the stock level is reduced (seemingly, at least), causing an error on the cart page:

### To Reproduce
1. Create a product that has stop level 1 set
![image](https://github.com/Automattic/woocommerce-payments/assets/133644769/5de4e2aa-8e4f-4482-bd23-9bcabad9bf9f)
2. Add this product to the cart
3. At the cart, click on Buy with WooPay
4. Click on the Cart button at the top left to go back to the site's cart

### Actual behavior
You're redirected to the cart and get a message:
"Sorry, we do not have enough “Stock 1” in stock to fulfill your order (0 available). We apologize for any inconvenience caused."
![image](https://github.com/Automattic/woocommerce-payments/assets/133644769/ca952896-a21a-44a0-8872-dc169337c9d2)
Going back to the product I can see that there's in fact 1 product in stock.

Opening the modal creates the draft order. So clicking the button and then closing the modal, or inputing a valid email in the email input and then closing it before inputing the OTP or getting redirected also causes this issue.

### Expected behavior
Don't make it seem like the stock level is affected by a not finished purchase.

### Additional context
#7046997-zen
Might be related to this: #7256

**Comment by csmcneill** [comment]
I can reproduce this, albeit inconsistently:

![NXZEL1.png](https://github.com/Automattic/woocommerce-payments/assets/82999806/c2aa2e60-f307-4b2d-8e4f-f4ff7943399a)

`?wc-ajax=wcpay_init_woopay` loads twice for me in my browser's Network tools. One is successful...

```
{
    "result": "success",
    "url": "https:\/\/pay.woo.com\/woopay\/?platform_checkout_key=[redacted]&user_session=[redacted]"
}
```

...and one isn't:

```
{"result":"failure","code":"invalid_request"}
```

**Comment by zmaglica** [comment]
This issue impacts WooPay, so assigning to team Heisenberg (based on team responsibilities Pc2DNy-3z-p2) @frosso . Assigning as part of Gamma Triage process PcreKM-yM-p2.

**Comment by c-shultz** [comment]
@bborman22 can we check if https://github.com/Automattic/woocommerce-payments/pull/7215 fixes this?

**Comment by bborman22** [comment]
I'm able to confirm that #7215 does **not** fix this issue. I believe the problem is caused by the timing / creation of the draft order, but it would require more digging.

**Comment by mouligreenlaw** [comment]
Hi there,
This ticket came back in the queue (7046997-zen) and and we haven't responded to them in 2 weeks I thought we should give them something.
Are there any updates to this issue?
Many thanks.

**Comment by LCmry** [comment]
From what I've seen so far, this happens because the stock gets put on hold for 10 minutes when a draft order is created, which occurs when the user starts the checkout process with WooPay.

<img width="272" alt="Screen Shot 2023-10-19 at 9 17 24 AM" src="https://github.com/Automattic/woocommerce-payments/assets/7651258/59cf4e92-d542-4bcb-9f55-dec7602782dc">


When I went to check out again through WooPay, this error occurred on the WooPay checkout:
`We're not able to process this request. Please refresh the page and try again.` So during that time period, it is not possible at this point to finish the checkout, which keeps the last item in stock in limbo.

**Comment by LCmry** [comment]
I think the holding stock might actually be for the duration that the associated order is in `draft` status, even if it should be release according to the setting. To my knowledge draft orders are deleted after 24 hours.

A couple questions for flows:
1. Why doesn't WooPay allow the checkout with the same cart and draft order? Should it?
2. Should we release stock if someone returns from the WooPay checkout page to the cart?

**Comment by c-shultz** [comment]
>Why doesn't WooPay allow the checkout with the same cart and draft order? Should it?

I would have expected that this would be allowed. @senadir, would this be something you could help us better understand? I think we'll need to dig into the WooPay side of things, but it would be helpful to double-check our understanding of how this should in a standard checkout flow using the Blocks/StoreAPI.

**Comment by senadir** [comment]
So on general, Cart doesn't hold stock generally. Once you start Checkout, stock is held for 10 minutes and this is tied to the Order id. When you place an order, the order status is changed to a stock reducing status (processing, on-hold, complete) and that would be the one responsible for reducing the stock (which is different from holding it).

Now, going from held to reduced in the same order should work fine, if you're creating another order or creating a new cart (not the same cart) then this might raise an issue.

WooPay ideally should resume from the same order (which can be hydrated from the session value `store_api_draft_order` but ideally if you're loading the same session between the 2 requests, then you should be using the same draft order.

**Comment by LCmry** [comment]
Summary from @bborman22 from our call:
- The issue with the stock reduction is that the call to `get_woopay_session` creates a draft order, particularly the `$preloaded_checkout_data checkout request in WooPay_Session::get_init_session_request`.
- This draft order puts that inventory on hold for 10 minutes.
- Then when loading the cart page, in `WC_Cart::check_cart_item_stock` the `$held_stock` is calculated and is supposed to take into account any quantities that are part of the current order. This is because we also count the `$required_stock` in the current cart, so we would be double counting it if we didn't exclude the current order from the held stock.
- The exact problem is that when calculating the held stock, it uses the `$current_session_order_id` which is pulled from `WC()->session->order_awaiting_payment` which in our tests during the call was always 0. We would expect it to be the draft order number instead.
- When manually setting the current session order ID to the draft order ID while debugging, the stock calculation works and no error is displayed.
- As a quick fix we tried `WC()->session->set( 'order_awaiting_payment', $checkout_data['order_id'] );` in the `get_init_session_request`, but this seemed to clear the cart data entirely, so it's likely that we're not doing that in the right way or at the right time to properly set it.

**Comment by LCmry** [comment]
Something I've also found is that this only affects the shortcode cart, things work fine going through the blocks cart. However, from what I can tell, `'order_awaiting_payment'` is still not set.

**Comment by ncsramona** [comment]
> that this only affects the shortcode cart

Now that's interesting! I didn't think there could be a difference, so I didn't test with the Block cart.

**Comment by bborman22** [comment]
So I dug in a bit on why there is a difference between shortcode and blocks cart and found that it's due to how it calculates the remaining quantity in the cart. On shortcode it uses `WC_Cart::check_cart_item_stock` and the process described above, mainly that it uses `WC()->session->order_awaiting_payment` to determine which order ID it's currently working with and exclude it from the held stock count. 

On the block checkout it uses `CartController::get_remaining_stock_for_product` which uses the draft order ID from `wc()->session->get( 'store_api_draft_order', 0 );`. Since WooPay uses the Store API this draft order is set after initializing a WooPay session, so blocks pages will be checking the right spot for the "current order ID" while shortcode won't.

They both ultimately are using the same underlying `ReserveStock::get_reserved_stock`, the only difference comes from what `$exclude_order_id` they pass in. Given that difference, I went to check on our previous quick fix idea of setting `order_awaiting_payment` when initializing the WooPay session. I found another example of it in the WC Subscriptions plugin `WCS_Cart_Renewal::set_order_awaiting_payment` and realized the cart was likely clearing because we weren't setting a new cart hash. I started experimenting a bit with that with no luck, but I think it might be the last piece.

```diff
diff --git a/includes/woopay/class-woopay-session.php b/includes/woopay/class-woopay-session.php
index 7a056e041..a947d4498 100644
--- a/includes/woopay/class-woopay-session.php
+++ b/includes/woopay/class-woopay-session.php
@@ -384,6 +384,15 @@ class WooPay_Session {
 		remove_filter( 'woocommerce_store_api_disable_nonce_check', '__return_true' );
 		$checkout_data = isset( $preloaded_checkout_data['/wc/store/v1/checkout'] ) ? $preloaded_checkout_data['/wc/store/v1/checkout']['body'] : '';
 
+		if ( ! empty( $checkout_data['order_id'] ) ) {
+			WC()->session->set( 'order_awaiting_payment', $checkout_data['order_id'] );
+			if ( is_callable( array( WC()->cart, 'get_cart_hash' ) ) ) {
+				$cart_hash = WC()->cart->get_cart_hash();
+				$order = wc_get_order( $checkout_data['order_id'] );
+				$order->set_cart_hash( $cart_hash );
+			}
+		}
+
 		$request = [
 			'wcpay_version'        => WCPAY_VERSION_NUMBER,
 			'user_id'              => $user->ID,
```

I think long term though, core WC shortcode cart should be checking for the `store_api_draft_order` AND the `order_awaiting_payment` when calculating the held stock.

**Comment by maxlaf** [comment]
7272424-zen

**Comment by LCmry** [comment]
I think I figured out why setting `order_awaiting_payment` in the session is not working. `WC_Cart::check_cart_item_stock` runs before `WooPay_Session::get_init_session_request` so it doesn't make a difference to set it there. A few questions remain:
1. How should this loading order be dealt with? 
     - It seems odd that this is the order, is this intended?
     - There is a filter `woocommerce_cart_item_required_stock_is_not_enough` that deals with the stock check that we might be able to use, but it would take care so we don't mess up needed checks and where to put it?
3. Why does `check_cart_item_stock` run first? It is my understanding that the session init is ajax run, so wouldn't that be sooner because the scripts are enqueued earlier? I'm not well-versed in this part though, so explanation would be very welcome.
4. Why did setting `order_awaiting_payment` empty the cart? @bborman22 I saw you were talking about the cart hash above, but I've looked at many instances of setting something in the session that don't need that. Wondering what it is about this timing or whatever else is affecting it that it may need the hash set?

**Comment by EjayhanFernandes** [comment]
7331050-zd-a8c - Same issue. I advised the following:

1. Disable WooPay under Payments > Settings > Express checkouts > Uncheck WooPay > Scroll to the bottom and Save changes.
2. Next, go to [WooCommerce > Orders > Drafts ](![0Zsag0.png](https://github.com/Automattic/woocommerce-payments/assets/81605775/95200c87-a794-4f46-8acd-c167b01041d3))> Move all draft orders [to the trash](![tmKWmS.gif](https://github.com/Automattic/woocommerce-payments/assets/81605775/414f3f1b-dc1d-432a-93d7-5b882efd6ee2)).
3. Wait until a fix is released on a new version release.

**Comment by frosso** [comment]
I wanted to cross-post the investigation @gpressutto5 did in this ticket, which seems related: https://github.com/Automattic/woocommerce-payments/issues/7256#issuecomment-1813245817

**Comment by LCmry** [comment]
Thanks @frosso !  @gpressutto5 do you know if #7764 resolves this issue as well? I don't have as much bandwidth right now to continue this, so if it's related feel free to address it.

**Comment by gpressutto5** [comment]
No, it won't resolve this issue. #7764 will prevent the draft orders from being created just by rendering the button, but when the modal opens, the draft order is still created, meaning closing the modal or clicking the Cart button at the top left to go back to the site's cart will still result in having stock issues.

**Comment by EjayhanFernandes** [comment]
Experienced in 7400415-zd-a8c. Told them to: 

1. Disable WooPay under Payments > Settings > Express checkouts > Uncheck WooPay > Scroll to the bottom and Save changes.
2. Next, go to WooCommerce > Orders > Drafts > Move all draft orders to the trash.
3. Wait until a fix is released on a new version release.

**Comment by csmcneill** [comment]
A member on the Woo Community Slack channel is affected by this issue: https://woocommercecommunity.slack.com/archives/C1KAZ91E3/p1701777564386919

**Comment by bborman22** [comment]
Quick update here, I've created an issue on core (https://github.com/woocommerce/woocommerce/issues/42784) to track this as it happens independent of WooPay and will have a PR up for it tomorrow to hopefully be included in WC 8.5.

**Comment by shaunkuschel** [comment]
Another report on 7469868-zen

**Comment by bborman22** [comment]
This fix was just merged into WC core and should be available in the WC 8.5.0 release scheduled for Jan 9th. I'll leave this issue open for now and track the fix through the WC beta and RC process.

**Comment by bborman22** [comment]
I was able to confirm that this issue is no longer present in WooCommerce 8.5.0.
